academy/ibc-dev/ibcapp_checkers.md

```diff
@@ -0,0 +1,342 @@
+
+```
+message MsgSendScore {
+  string sender = 1;
+  string port = 2;
+  string channelID = 3;
+  uint64 timeoutTimestamp = 4;
+}
+```
+
+You will also need to adjust the `x/checkers/client/cli/tx_score.go` to make it consume no arguments because the `sender` can only be the signer of the transaction and the other fields of the `ScorePacketData` are fetched by the `message_server_score.go`. 
+
+Now we can send such a score packet, what do we do if we receive one? We can implement our logic in the `x/checkers/keeper/score.go`:
+
+```golang
+
+...
+
+// OnRecvScorePacket processes packet reception
+func (k Keeper) OnRecvScorePacket(ctx sdk.Context, packet channeltypes.Packet, data types.ScorePacketData) (packetAck types.ScorePacketAck, err error) {
+  // validate packet data upon receiving
+  if err := data.ValidateBasic(); err != nil {
+    return packetAck, err
+  }
+
+  score:= types.WinningPlayer {
+    PlayerAddress: data.PlayerAddress,
+    WonCount: data.WonCount,
+    DateAdded: data.DateAdded,
+  }
+
+  leaderboard, found := k.GetLeaderboard(ctx)
+  
+  if !found {
+    panic("Leaderboard not found")
+  }
+
+  found_in_leaderboard := false
+  for i := range leaderboard.Winners {
+      if leaderboard.Winners[i].PlayerAddress == data.PlayerAddress {
+          // found the player, update the fields
+          leaderboard.Winners[i].WonCount = data.WonCount
+          leaderboard.Winners[i].DateAdded = data.DateAdded
+          found_in_leaderboard = true
+          break
+      }
+  }
+
+  // we cannot find the player in the leaderboard
+  if(!found_in_leaderboard) {
+    updated:= append(leaderboard.Winners, &score)
+    leaderboard.Winners = updated
+  } 
+    
+  k.SetLeaderboard(ctx, leaderboard)
+
+  return packetAck, nil
+}
+
+...
+
+```
+
+Above you can see that we differentiate between two cases: either the player is not in the leaderboard yet so we need to append the player and the score or the player can be found in the leaderboard and we need to update the player state. That is it basically. Now the `checkers` blockchain will allow you to send and receive score packets. You can add some additional logic, e.g. check if the packet is valid in the `x/checkers/types/packet_score.go`:
+
+```golang
+...
+
+// ValidateBasic is used for validating the packet
+func (p ScorePacketData) ValidateBasic() error {
+
+  // TODO: Validate the packet data
+
+  // return error if player address is empty
+  if p.PlayerAddress == "" {
+    return errors.New("Player address cannot be empty")
+  }
+
+  return nil
+}
+
+...
+```
+
+## Adjust `leaderboard`
+
+In the `leaderboard` chain, you need first to do the same steps as in the [Migration](add_link) section for the `checkers` chain and include a leaderboard but you will need a slightly different leaderboard structure for the `leaderboard` chain:
+
+```
+message Board {
+  repeated PlayerInfo player = 1; 
```

Comment:
In Checkers, `PlayerInfo` refers to something other than how it is defined here. Perhaps we want to prevent confusion and rename it there to `WinnerInfo` or `PlayerInLeaderboard`.